python/dgl/nn/pytorch/softmax.py

```diff
@@ -0,0 +1,104 @@
+"""Torch modules for graph related softmax."""
+# pylint: disable= no-member, arguments-differ
+import torch as th
+from torch import nn
+
+from ... import function as fn
+
+__all__ = ['EdgeSoftmax']
+
+class EdgeSoftmax(nn.Module):
+    r"""Apply softmax over signals of incoming edges.
+
+    For a node :math:`i`, edgesoftmax is an operation of computing
+
+    .. math::
+      a_{ij} = \frac{\exp(z_{ij})}{\sum_{j\in\mathcal{N}(i)}\exp(z_{ij})}
+
+    where :math:`z_{ij}` is a signal of edge :math:`j\rightarrow i`, also
+    called logits in the context of softmax. :math:`\mathcal{N}(i)` is
+    the set of nodes that have an edge to :math:`i`.
+
+    An example of using edgesoftmax is in
+    `Graph Attention Network <https://arxiv.org/pdf/1710.10903.pdf>`__ where
+    the attention weights are computed with such an edgesoftmax operation.
+
+    Parameters
+    ----------
+    logits_name : str
+        A field name such that g.edata[logits_name] gives the logits :math:`z_{ij}`.
+        based on which edgesoftmax will be computed.
+    max_logits_name : str
+        A field name such that g.ndata[max_logits_name] will store
+        :math:`\max_{j\in\mathcal{N}(i)}z_{ij}`.
+    normalizer_name : str
+        A field name such that g.ndata[normalizer_name] will store
+        :math:`\sum_{j\in\mathcal{N}(i)}exp(z_{ij})`
+    """
+    def __init__(self,
+                 logits_name,
+                 max_logits_name,
+                 normalizer_name):
+        super(EdgeSoftmax, self).__init__()
+        # compute the softmax
+        self._logits_name = logits_name
+        self._max_logits_name = max_logits_name
+        self._normalizer_name = normalizer_name
+
+    def forward(self, logits, graph):
+        r"""Compute edge softmax.
+
+        Parameters
+        ----------
+        logits : torch.Tensor
+            The input edge feature
+        graph : DGLGraph
+            The graph.
+
+        Returns
+        -------
+        Unnormalized scores : torch.Tensor
+            This part gives :math:`\exp(z_{ij})`'s
+        Normalizer : torch.Tensor
+            This part gives :math:`\sum_{j\in\mathcal{N}(i)}\exp(z_{ij})`
```

Comment:
I guess eventually we want to just return the normalized scores as normal softmax does. But I'm fine with this right now. We could change this when we have our kernel to do that more easily.